pkg/pmem-csi-driver/controllerserver-master.go

```diff
@@ -75,17 +74,65 @@ func NewMasterControllerServer(rs *registryserver.RegistryServer) *masterControl
 		csi.ControllerServiceCapability_RPC_LIST_VOLUMES,
 		csi.ControllerServiceCapability_RPC_GET_CAPACITY,
 	}
-	return &masterController{
+	cs := &masterController{
 		DefaultControllerServer: NewDefaultControllerServer(serverCaps),
 		rs:                      rs,
 		pmemVolumes:             map[string]*pmemVolume{},
 	}
+
+	rs.AddListener(cs)
+
+	return cs
 }
 
 func (cs *masterController) RegisterService(rpcServer *grpc.Server) {
 	csi.RegisterControllerServer(rpcServer, cs)
 }
 
+// OnNodeAdded retrieves the existing volumes at recently added Node.
+// It uses ControllerServer.ListVolume() CSI call to retrieve volumes.
+func (cs *masterController) OnNodeAdded(ctx context.Context, node *registryserver.NodeInfo) {
+	conn, err := cs.rs.ConnectToNodeController(node.NodeID)
+	if err != nil {
+		glog.Warningf("Failed to connect to node controller at : %s on node %s: %s", node.Endpoint, node.NodeID, err.Error())
+		return
+	}
+
+	csiClient := csi.NewControllerClient(conn)
+	resp, err := csiClient.ListVolumes(ctx, &csi.ListVolumesRequest{})
+	if err != nil {
+		glog.Warningf("Failed to get volumes on node %s: %s", node.NodeID, err.Error())
+	}
```

Comment:
I'm a bit concerned about the lack of error handling here. If anything goes wrong, all we get are some warnings in the controller log.

No admin is going to find them there, if they even know that they need to look and what they mean.

There's no recovery mechanism. Even if an admin had some way to fix the situation, having to do so manually is not how Kubernetes is meant to work.

Let's fix this after merging this PR. I'll file an issue.